### PR TITLE
fix: symlinks in directory targets

### PR DESCRIPTION
**Pull request by rgrinberg**
compute digests for symlinks present inside directory targets

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

ps-id: a0b24bf1-b7da-413f-a2d6-08f977cc8973

### CODE REVIEW

**Comment by ejgallego** [comment]
Tested on our setup, works fine! Thanks a lot @rgrinberg

**Comment by ejgallego** [comment]
> Tested on our setup, works fine! Thanks a lot @rgrinberg

Actually there seems to be a problem with the rule still, the directory is always generated, even if the deps didn't change.

**Comment by ejgallego** [comment]
> > Tested on our setup, works fine! Thanks a lot @rgrinberg
> 
> Actually there seems to be a problem with the rule still, the directory is always generated, even if the deps didn't change.

Ha, never mind! This was a problem in our end, we were doing some quite questionable, we both had:

- `node_modules` as a directory target
- `(dirs :standard \ node_modules)`

This prevented dune to cache `node_modules` among rebuilds. We'll need to figure out what to do here, as it is a common use case for us to want to have a `node_modules` in the build dir sometimes.

**Comment by Alizter** [comment]
@ejgallego Wouldn't that only matter if you promoted the directory target?

**Comment by ejgallego** [comment]
> @ejgallego Wouldn't that only matter if you promoted the directory target?

Yes, we were doing "manual promotion", that is say, calling `npm` directly. That worked out fine as dune didn't see the target conflict at all. We can indeed think of promoting now that the regular directory target works, but we have other issues to sort out such as `webpack` actually modifying `node_modules`....

File: src/dune_digest/dune_digest.ml [hunk 1]
```diff
@@ -112,43 +116,59 @@ exception
...
 
-let rec path_with_stats ~allow_dirs path (stats : Stats_for_digest.t) :
+let path_with_stats ~allow_dirs path (stats : Stats_for_digest.t) :
     Path_digest_result.t =
+  let rec loop path (stats : Stats_for_digest.t) =
+    match stats.st_kind with
+    | S_LNK ->
+      let executable =
+        Path.Permissions.test Path.Permissions.execute stats.st_perm
+      in
```
Comment (Alizter) [inline src/dune_digest/dune_digest.ml:126]: You can lift this out of the match and share it with the other branch. Maybe thunk it if you don't want it evaluated?

File: src/dune_digest/dune_digest.ml [hunk 2]
```diff
@@ -112,43 +116,59 @@ exception
...
+      Dune_filesystem_stubs.Unix_error.Detailed.catch
+        (fun path ->
+          let contents = Unix.readlink (Path.to_string path) in
+          path_with_executable_bit ~executable ~content_digest:contents)
```
Comment (Alizter) [inline src/dune_digest/dune_digest.ml:130]: Maybe make a `link_with_executable_bit` so that this is easier to read?

**Review by Alizter — APPROVED** [review]
LGTM, two suggestions

**Comment by Alizter** [comment]
I suppose if you bunch all your actions into one action then the directory target would work. But I think what is missing is a way to compose directory target actions. That might not be a good or easy idea.

File: src/dune_digest/dune_digest.ml [hunk 3]
```diff
@@ -66,13 +66,17 @@ let generic a =
   Metrics.Timer.record "generic_digest" ~f:(fun () ->
       string (Marshal.to_string a [ No_sharing ]))
 
-let file_with_executable_bit ~executable path =
+let path_with_executable_bit =
```
Comment (snowleopard) [inline src/dune_digest/dune_digest.ml:69]: What is the point of this renaming? It seems like if we speak of executable bits, we're speaking of files.

**Review by snowleopard — COMMENTED** [review]
Thanks! Could you also add a test that checks that we digest symbolic links correctly (as opposed to merely checking that we don't refuse to digest them)?

**Comment by rgrinberg** [comment]
> Thanks! Could you also add a test that checks that we digest symbolic links correctly (as opposed to merely checking that we don't refuse to digest them)?

Will do. Looks we already have tests but they're accidentally disabled. Will enable them in a different PR first.

**Comment by rgrinberg** [comment]
@snowleopard okay, the PR now demonstrates the unit tests for directory digests are passing.

File: src/dune_digest/dune_digest.ml (hunk 3 above)
Comment (rgrinberg) [reply src/dune_digest/dune_digest.ml:69]: Directories also have executable bits though. And on macos and freebsd, so can symlinks.

Comment (snowleopard) [reply src/dune_digest/dune_digest.ml:69]: Right. Thanks, makes sense.

**Comment by snowleopard** [comment]
> @snowleopard okay, the PR now demonstrates the unit tests for directory digests are passing.

Thanks! That's good but I think our code coverage for this feature is still poor. We don't test this line, for example:

```ocaml
let contents = Unix.readlink (Path.to_string path) in
```

That is, you could do a wrong thing here (e.g. digest the contents of the linked file instead) and the current test would still pass. I think it woud be nice to have something more rigorous, like building a rule with a symlink target, making some changes, and testing that correct rules are being rebuilt.

**Comment by emillon** [comment]
Thanks. I'm merging this for now in order to release 3.5 but the extra tests can be added in 3.6.

**Comment by ejgallego** [comment]
Thanks a lot folks this PR is great and we should be able to test directory targets a lot more now !

**Pull request merged**